packages/truffle-core/lib/environment.js

```diff
@@ -60,16 +60,14 @@ const Environment = {
 
     async function detectNetworkId() {
       const providerNetworkId = await web3.eth.net.getId();
-      if (network_id != "*") {
+      if (network_id !== "*") {
         // Ensure the network id matches the one in the config for safety
-        if (providerNetworkId !== network_id) {
-          const error = {
-            message:
-              `The network id specified in the truffle config ` +
-              `(${network_id}) does not match the one returned by the network ` +
-              `(${providerNetworkId}).  Ensure that both the network and the ` +
-              `provider are properly configured.`
-          };
+        if (JSON.stringify(providerNetworkId) !== JSON.stringify(network_id)) {
```

Comment:
We should just use `.toString()` here. We want the loss of precision. Also, if `providerNetworkId` is a BN then we won't be able to JSONify that